src/networks.jl

```diff
@@ -994,3 +994,262 @@ Carries on the spatial convolution respecting periodicity at the boundaries.
 function (layer::CircularConv)(x)
     layer.conv(NNlib.pad_circular(x, layer.pad))
 end
+
+"""
+    CliMAgen.ControlNet
+
+The struct containing the parameters and layers of the ControlNet architecture.
+"""
+struct ControlNet{N}
+    net::N
+    trainable::Bool  # whether the network is trainable
+end
+
+"""
+    ControlNet(net::N; trainable::Bool=false)
+
+Creates a ControlNet with the given neural network `net` and whether it is trainable.
+"""
+function ControlNet(net::N; trainable::Bool=false) where N
+    return ControlNet{N}(net, trainable)
+end
+
+@functor ControlNet
+
+"""
+    (net::ControlNet)(x)
+
+Evaluates the neural network of the ControlNet model on `x`.
+"""
+function (c::ControlNet)(x)
+    return c.net(x)
+end
+
+"""
+    Flux.params(::ControlNet)
+Returns the trainable parameters of the ControlNet).
+"""
+Flux.params(c::ControlNet) = c.trainable ? Flux.params(c.net) : nothing
+
+"""
+    CliMAgen.ControlledNoiseConditionalScoreNetwork
+
+The struct containing the parameters and layers
+of the Noise Conditional Score Network architecture,
+with the option to include a mean-bypass layer.
+
+# References
+Unet: https://arxiv.org/abs/1505.04597
+"""
+struct ControlledNoiseConditionalScoreNetwork{N}
+    "The layers of the network"
+    layers::NamedTuple
+    "A control network to condition the output of the U-net"
+    control_net::N
```

Comment:
It seems like the output of the control net needs to be a vector of length embed_dim (same as for the time embedding). Is this correct? Since the control net is made exterior to the NCSN, where embed_dim is defined, we  should maybe just add some documentation stating this, or maybe an @assert statement somewhere, if we can?